device/sycl/src/clusterization/clusterization_algorithm.cpp

```diff
@@ -55,93 +60,113 @@ clusterization_algorithm::output_type clusterization_algorithm::operator()(
+
+    // Vector with numbers of found clusters in each module, later transformed
+    // into prefix sum vector
+    vecmem::data::vector_buffer<std::size_t> clusters_per_module_prefix(
+        num_modules, m_mr.main);
+    m_copy->setup(clusters_per_module_prefix);
+
+    // Get the prefix sum of the cells and copy it to the device buffer
+    const device::prefix_sum_t cells_prefix_sum = device::get_prefix_sum(
+        cell_sizes, (m_mr.host ? *(m_mr.host) : m_mr.main));
+    vecmem::data::vector_buffer<device::prefix_sum_element_t>
+        cells_prefix_sum_buff(cells_prefix_sum.size(), m_mr.main);
+    m_copy->setup(cells_prefix_sum_buff);
+    (*m_copy)(vecmem::get_data(cells_prefix_sum), cells_prefix_sum_buff);
 
     // Clusters sum kernel
-    traccc::sycl::clusters_sum(cells_view, sparse_ccl_indices, *total_clusters,
-                               cluster_prefix_sum, clusters_per_module,
-                               m_queue);
+    traccc::sycl::clusters_sum(cells_data, sparse_ccl_indices,
+                               clusters_per_module_prefix, m_queue);
+
+    // Copy the sizes of clusters per each module to the host
+    vecmem::vector<std::size_t> clusters_per_module_host(
+        m_mr.host ? m_mr.host : &(m_mr.main));
+    (*m_copy)(clusters_per_module_prefix, clusters_per_module_host);
+
+    // Create the measurements and spacepoints buffer already here
+    // in order not to use additional memory space
+
+    // Resizable buffer for the measurements
+    measurement_container_types::buffer measurements_buffer{
+        {num_modules, m_mr.main},
+        {std::vector<std::size_t>(num_modules, 0),
+         std::vector<std::size_t>(clusters_per_module_host.begin(),
+                                  clusters_per_module_host.end()),
+         m_mr.main, m_mr.host}};
+    m_copy->setup(measurements_buffer.headers);
+    m_copy->setup(measurements_buffer.items);
+
+    // Spacepoint container buffer to fill in spacepoint formation
+    spacepoint_container_types::buffer spacepoints_buffer{
+        {num_modules, m_mr.main},
+        {std::vector<std::size_t>(num_modules, 0),
+         std::vector<std::size_t>(clusters_per_module_host.begin(),
+                                  clusters_per_module_host.end()),
+         m_mr.main, m_mr.host}};
+    m_copy->setup(spacepoints_buffer.headers);
+    m_copy->setup(spacepoints_buffer.items);
+
+    // Perform the exclusive scan operation. It will results with a vector of
+    // prefix sums, starting with 0, until the second to last sum. The total
+    // number of clusters in this case will be the last prefix sum + the
+    // clusters in the last idx
+    unsigned int total_clusters = clusters_per_module_host.back();
+    std::exclusive_scan(clusters_per_module_host.begin(),
+                        clusters_per_module_host.end(),
+                        clusters_per_module_host.begin(), 0);
+    total_clusters += clusters_per_module_host.back();
```

Comment:
Wait... :confused: Do I understand correctly that you're doing the following in a much more complicated way here?

```c++
const std::size_t total_clusters = std::accumulate(clusters_per_module_host.begin(), clusters_per_module_host.end(), 0);
```

I had to look at the description of [std::exclusive_scan](https://en.cppreference.com/w/cpp/algorithm/exclusive_scan) for a bit to see what you're doing here...

Reply:
Partially. I need the `total_clusters` but the main reason for this operation is to get a vector of prefix sums, starting from 0 to up to second-to-last prefix sum. So after `exclusive scan`, the `clusters_per_module_host` is this vector. It is then copied to the device, into `cl_per_module_prefix_buff`

Reply:
He uses this prefix sum later as well, so in that case it's a pretty clever way to combine the prefix sum and accumulation operations.

Reply:
Yes, exactly, I could use std accumulate alongside the exclusive scan but that's adding some more work which is avoided in this way.

Reply:
Okay, let's go with this then.